mongodbatlas/fw_resource_mongodbatlas_search_deployment_model_test.go

```diff
@@ -0,0 +1,108 @@
+package mongodbatlas
+
+import (
+	"context"
+	"reflect"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-framework/types/basetypes"
+	"go.mongodb.org/atlas-sdk/v20231115001/admin"
+)
+
+type sdkToTFModelTestCase struct {
+	SDKResp         *admin.ApiSearchDeploymentResponse
+	expectedTFModel *tfSearchDeploymentRSModel
+	name            string
+	clusterName     string
+}
+
+func TestSearchDeploymentSDKToTFModel(t *testing.T) {
+	testCases := []sdkToTFModelTestCase{
+		{
+			name:        "Complete SDK response",
+			clusterName: "Cluster0",
+			SDKResp: &admin.ApiSearchDeploymentResponse{
+				Id:        admin.PtrString("111111111111111111111111"),
+				GroupId:   admin.PtrString("222222222222222222222222"),
+				StateName: admin.PtrString("IDLE"),
+				Specs: []admin.ApiSearchDeploymentSpec{
+					{
+						InstanceSize: "S20_HIGHCPU_NVME",
+						NodeCount:    2,
+					},
+				},
+			},
+			expectedTFModel: &tfSearchDeploymentRSModel{
+				ID:          types.StringValue("111111111111111111111111"),
+				ClusterName: types.StringValue("Cluster0"),
+				ProjectID:   types.StringValue("222222222222222222222222"),
+				StateName:   types.StringValue("IDLE"),
+				Specs:       tfSpecsList(t, "S20_HIGHCPU_NVME", 2),
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			resultModel, diags := newTFSearchDeployment(context.Background(), tc.clusterName, tc.SDKResp, nil)
+			if diags.HasError() {
+				t.Errorf("Unexpected errors found")
+			}
+			if !reflect.DeepEqual(resultModel, tc.expectedTFModel) {
+				t.Errorf("created terraform model did not match expected output")
+			}
+		})
+	}
+}
+
+type tfToSDKModelTestCase struct {
+	name           string
+	tfModel        *tfSearchDeploymentRSModel
+	expectedSDKReq admin.ApiSearchDeploymentRequest
+}
+
+func TestSearchDeploymentTFModelToSDK(t *testing.T) {
+	testCases := []tfToSDKModelTestCase{
+		{
+			name: "Complete TF state",
+			tfModel: &tfSearchDeploymentRSModel{
+				ID:          types.StringValue("111111111111111111111111"),
+				ClusterName: types.StringValue("Cluster0"),
+				ProjectID:   types.StringValue("222222222222222222222222"),
+				StateName:   types.StringValue("IDLE"),
+				Specs:       tfSpecsList(t, "S20_HIGHCPU_NVME", 2),
+			},
+			expectedSDKReq: admin.ApiSearchDeploymentRequest{
+				Specs: []admin.ApiSearchDeploymentSpec{
+					{
+						InstanceSize: "S20_HIGHCPU_NVME",
+						NodeCount:    2,
+					},
+				},
+			},
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			apiReqResult := newSearchDeploymentReq(context.Background(), tc.tfModel)
+			if !reflect.DeepEqual(apiReqResult, tc.expectedSDKReq) {
+				t.Errorf("created sdk model did not match expected output")
+			}
+		})
+	}
+}
+
+func tfSpecsList(t *testing.T, instanceSize string, nodeCount int64) basetypes.ListValue {
```

Comment:
nit: we don't have a team guideline hence so far it's up to engineer's preference. My suggestion is to avoid using separate method to "optimize" code in unit testing. Like Google does, we should maximize in-test context without deferring to separate methods.

Reply:
you don't have to follow this, I just wanted to bring this up.

Reply:
I see your point. I extracted this method mainly for readability of the test case:
```
expectedTFModel: &tfSearchDeploymentRSModel{
  ID:          types.StringValue("111111111111111111111111"),
  ClusterName: types.StringValue("Cluster0"),
  ProjectID:   types.StringValue("222222222222222222222222"),
  StateName:   types.StringValue("IDLE"),
  Specs:       tfSpecsList(t, "S20_HIGHCPU_NVME", 2),
}
```
Unfortunately there is no way to define a terraform `types.List` inlined.